libsel4vm/src/boot.c

```diff
@@ -88,10 +88,17 @@ int vm_assign_vcpu_target(vm_vcpu_t *vcpu, int target_cpu)
         return -1;
     }
     vm_vcpu_t *target_vcpu = vm_vcpu_for_target_cpu(vcpu->vm, target_cpu);
-    if (target_vcpu) {
+    if (target_vcpu && (target_vcpu != vcpu)) {
         ZF_LOGE("Failed to assign target cpu - A VCPU is already assigned to core %d", target_cpu);
         return -1;
     }
+
+#if CONFIG_MAX_NUM_NODES > 1
+    if (seL4_TCB_SetAffinity(vcpu->tcb.tcb.cptr, target_cpu)) {
```

Comment:
Nitpicking: this just moves the check, but the logic is remains semantically broken here. It's easier to understand if this would be written as: 

```suggestion
    if (0 != seL4_TCB_SetAffinity(vcpu->tcb.tcb.cptr, target_cpu)) {
```
Or even more verbosely with an error mesage:
```
    int err = seL4_TCB_SetAffinity(vcpu->tcb.tcb.cptr, target_cpu);
    if (err) {
        ZF_LOGE("[vCPU %u] Failed to set CPU affinity to %d", target_cpu);
   }
```

Reply:
Done